### PR TITLE
Update JSCallResultTypeHelper.cs

### PR DESCRIPTION
**Pull request by MackinnonBuck**
#### Description

Currently, each .NET->JavaScript interop call results in `JSCallResultTypeHelper.FromGeneric<TResult>()` comparing `TResult` with up to three `IJSObjectReference` types. We can optimize this slightly for the general case by first checking if `typeof(TResult).Assembly` matches the assembly containing `JSCallResultType` before performing the other checks.

#### Customer Impact

Customers may see a slight performance improvement for JS interop calls.

#### Regression?

No.

#### Risk

Very small risk. This is a small change and existing tests cover this area.

Addresses #25624

### CODE REVIEW

**Review by dougbu — DISMISSED** [review]
This PR fixes a build break. We need it in ASAP

**Comment by dougbu** [comment]
@Pilchie should I proactively merge this to unbreak the build, letting @SteveSandersonMS and @MackinnonBuck follow up on the C# change if there's something wrong there❔

**Comment by halter73** [comment]
Can this be tested?

File: src/Shared/JSInterop/JSCallResultTypeHelper.cs [hunk 1]
```diff
@@ -1,15 +1,37 @@
...
+            {
+                if (resultType == typeof(IJSObjectReference)
+                    || resultType == typeof(IJSInProcessObjectReference)
+                    || resultType == typeof(IJSUnmarshalledObjectReference))
+                {
+                    return JSCallResultType.JSObjectReference;
+                }
+                else
+                {
+                    throw new ArgumentException($"JS interop cannot supply an instance of type '{resultType}'.");
```
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs:28]: TBH I’d prefer not to list the types. The set of useful ones varies depending on the host environment.

[the remark this answers is not in the document]
Comment (MackinnonBuck) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs]: @SteveSandersonMS I've modified the message to suggest using `IJSObjectReference`, without listing the others for the reason you mentioned. Do you think that's fine, or do you prefer we don't list any types at all?

**Comment by Pilchie** [comment]
@MackinnonBuck please ping when CI is green and this is signed off.  Dismissing @dougbu's signoff since I presume it was based on *just* the build fix.

**Review by SteveSandersonMS — APPROVED** [review]
Great!

File: src/Shared/JSInterop/JSCallResultTypeHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs]: Looks good like this

File: src/Shared/JSInterop/JSCallResultTypeHelper.cs [hunk 2]
```diff
@@ -1,15 +1,39 @@
...
+            {
+                if (resultType == typeof(IJSObjectReference)
+                    || resultType == typeof(IJSInProcessObjectReference)
+                    || resultType == typeof(IJSUnmarshalledObjectReference))
+                {
+                    return JSCallResultType.JSObjectReference;
+                }
+                else
+                {
+                    throw new ArgumentException(
```
Comment (pranavkm) [inline src/Shared/JSInterop/JSCallResultTypeHelper.cs:28]: Wouldn't this trip on `DotNetObjectReference`? Should we check if the type derives from `IJSObjectReference`?

Comment (MackinnonBuck) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs:28]: Good catch. Looks like the current tests didn't catch this because they wrap `DotNetObjectReference` in a struct, so the generic parameter bypasses these checks.

Another thing I've realized is that if the customer creates their own class deriving from `IJSObjectReference` and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

It seems that for the exception to be thrown if and only if the user is asking for an `IJSObjectReference` type that is not one of the three allowed types, we would have to:
1. Check if the generic parameter type can be assigned to `IJSObjectReference` (something like `typeof(IJSObjectReference).IsAssignableFrom(resultType)`).
2. Check if the generic parameter type does not match one of the three accepted types.

But at that point, I'm not sure the new solution is really better than we we already had (@SteveSandersonMS expressed performance concerns with `Type.IsAssignableFrom` in a previous PR).

Thoughts?

Comment (pranavkm) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs:28]: > Another thing I've realized is that if the customer creates their own class deriving from IJSObjectReference and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

I think that should be fine. If they're dedicated enough to subtype this, we'll let them manage it's use. Could we simply check for the presence of the types we know users could accidentally misuse - `JSObjectReference` , and `JSInProcessObjectReference` and throw?

Comment (SteveSandersonMS) [reply src/Shared/JSInterop/JSCallResultTypeHelper.cs:28]: > Another thing I've realized is that if the customer creates their own class deriving from IJSObjectReference and attempts to grab an instance of it via JS interop, the new exception will not be thrown since their own type is from a different assembly. Should we handle this case as well?

That doesn't seem like a problem to me either. We don't need to recognize 3rd-party subtypes of `IJSObjectReference` since there's no way we'd know which concrete instance type to create. So I'd expect it to be treated as `JSCallResultType.Default`, as it already is here.

Good catch on `DotNetObjectReference` though! To be honest, I'm not massively bothered about providing a specific exception if people try to use `JSObjectReference` instead of `IJSObjectReference`. The equivalent would be trying to `@inject JSRuntime JS` instead of `IJSRuntime`, and we don't have any exception for that except the default one from DI.

I think if someone does mistakenly use `InvokeAsync<JSObjectReference>(...)` then they'll get an exception from the JSON deserializer saying it doesn't have a default constructor, which is basically good enough because then the developer will have no choice but to realise they are doing it wrong and will have to check docs again or another working implementation to copy/paste.

So in summary I'd be fine with a simplified update like:

```cs
if ((resultType.Assembly == _currentAssembly)
  && (resultType is one of the three known IJSObjectReference ones))
{
    return JSCallResultType.JSObjectReference;
}
else
{
    return JSCallResultType.Default;
}
```

We could even choose to leave the previous implementation as-is, but adding the `resultType.Assembly == _currentAssembly` check seems at least marginally useful.

**Comment by Pilchie** [comment]
Approved for .NET 5.0 RC2.
